### ISSUE TITLE
mod_auth_cas v1.2 w/ CentOS 7, httpd 2.4, APR 1.7 - undefined symbol: cas_attr_builder_new

### ISSUE THREAD

**Issue by rbargerhuff** [apereo/mod_auth_cas#197]
Hello, I have been trying to get mod_auth_cas v1.2 to work in our work's base CentOS 7 environment. I am very close but I cannot get past the following error when starting apache:

`httpd: Syntax error on line 56 of /etc/httpd/conf/httpd.conf: Syntax error on line 8 of /etc/httpd/conf.modules.d/10-auth_cas.conf: Cannot load modules/mod_auth_cas.so into server: /etc/httpd/modules/mod_auth_cas.so: undefined symbol: cas_attr_builder_new
`

I followed the instructions that @dhawes outlined in the following post:
https://github.com/apereo/mod_auth_cas/issues/173#issuecomment-507773213

From what I can tell, everything compiled from source correctly. I hit a snag which I overcame when trying to get _httpd_ to use APR v1.7.

Configuring the LD_LIBRARY_PATH in /etc/sysconfig/httpd would not work. I instead needed to do the following:

1. Created following file: /etc/ld.so.conf.d/apr-1.7.x.conf with following content: /usr/local/apr/lib
2. Ran` ldconfig -V` as root.
3. `httpd -V` confirmed the following output:

```
[root@localhost ld.so.conf.d]# httpd -V
Server version: Apache/2.4.6 (CentOS)
Server built:   Apr  2 2020 13:13:23
Server's Module Magic Number: 20120211:24
Server loaded:  APR 1.7.0, APR-UTIL 1.5.2
Compiled using: APR 1.4.8, APR-UTIL 1.5.2
```

I then performed the following:
**Created** the following file: /etc/httpd/conf.modules.d/10-auth_cas.conf with the following content:

```
# mod_auth_cas is an Apache 2.2/2.4 compliant module that supports the
# CASv1 and CASv2 protocols
#
<IfModule !ssl_module>
    LoadModule ssl_module modules/mod_ssl.so
</IfModule>

LoadModule auth_cas_module modules/mod_auth_cas.so
```

**Created** following file: /etc/httpd/conf.d/auth_cas.conf with following content:

```
# Configuration for mod_auth_cas
#
# For more information please visit
#
#    https://github.com/apereo/mod_auth_cas
#
# Or reference /usr/share/doc/mod_auth_cas*/README
#

CASCookiePath /var/cache/httpd/mod_auth_cas/
CASLoginURL https://login.example.org/cas/login
CASValidateURL https://login.example.org/cas/serviceValidate

#Protect a "Location" or "Directory" block in your Apache
#  configuration:
#
#       <Location /secured>
#               Authtype CAS
#               require valid-user
#       </Location>

#If SAML-delivered attribute authorization is also desired, use the
#  samlValidate URL, enable SAML validation, and specify cas-attribute
#  in your require rule (please note: both attribute name and value are
#  case-sensitive):
#
#    CASValidateSAML On
#
#    <Location /secured>
#               Authtype CAS
#        require cas-attribute edupersonaffiliation:staff
#     </Location>

```

**Installed** mod_ssl using the following command:
`yum install mod_ssl`

After everything in place, and making sure everything is configured correctly for the most part, httpd will not start:

`httpd: Syntax error on line 56 of /etc/httpd/conf/httpd.conf: Syntax error on line 8 of /etc/httpd/conf.modules.d/10-auth_cas.conf: Cannot load modules/mod_auth_cas.so into server: /etc/httpd/modules/mod_auth_cas.so: undefined symbol: cas_attr_builder_new
`

My CentOS 7 environment meets all the requirements detailed in the read me for mod_auth_cas.

Can anyone offer any assistance?

**Comment by dhawes** [comment]
What do you get for:

nm /etc/httpd/modules/mod_auth_cas.so | grep cas_attr_builder_new

I tried this again in a CentOS7 docker image and httpd started without issue.

**Comment by agilis1** [comment]
When I execute that command I am returned the following:

`                 U cas_attr_builder_new
`

**Comment by dhawes** [comment]
Is it the same for your src/.libs/mod_auth_cas.so where you compiled mod_auth_cas?

**Comment by agilis1** [comment]
Performing the same command...

```
vagrant@localhost /opt/mod_auth_cas/src/.libs (master *=) $ nm mod_auth_cas.so
```

```
0000000000214068 B __bss_start
                 U cas_attr_builder_add
                 U cas_attr_builder_new

```

Also, for your reference... `uname -a:`
`Linux localhost.localdomain 3.10.0-1127.19.1.el7.x86_64 #1 SMP Tue Aug 25 17:23:54 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux`

Thank you for helping. It is greatly appreciated.

**Comment by agilis1** [comment]
More for your reference:

Compiling apr-1.7 from source:
`sudo ./configure && sudo make install`

Compiling mod_auth_cas from source: 
`sudo autoreconf -ivf && CPPFLAGS=-I/usr/local/apr/include/apr-1 LDFLAGS=-L/usr/local/apr/lib ./configure && sudo make && sudo make install
`

**Comment by dhawes** [comment]
That's odd that those symbols are undefined. Can you post the output of your make command?

**Comment by agilis1** [comment]
I absolutely can.

```vagrant@localhost /opt/mod_auth_cas (master *=) $ sudo autoreconf -ivf && CPPFLAGS=-I/usr/local/apr/include/apr-1 LDFLAGS=-L/usr/local/apr/lib ./configure && sudo make && sudo make install
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force -I m4
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in `.'.
libtoolize: copying file `./ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIR, `m4'.
libtoolize: copying file `m4/libtool.m4'
libtoolize: copying file `m4/ltoptions.m4'
libtoolize: copying file `m4/ltsugar.m4'
libtoolize: copying file `m4/ltversion.m4'
libtoolize: copying file `m4/lt~obsolete.m4'
autoreconf: running: /usr/bin/autoconf --force
autoreconf: running: /usr/bin/autoheader --force
autoreconf: running: automake --add-missing --copy --force-missing
autoreconf: Leaving directory `.'
checking whether to enable maintainer-specific portions of Makefiles... no
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking for library containing strerror... none required
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking for style of include used by make... GNU
checking whether make supports nested variables... yes
checking dependency style of gcc... gcc3
checking whether make supports nested variables... (cached) yes
checking for gcc... (cached) gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking whether gcc and cc understand -c and -o together... yes
checking whether C compiler accepts -Wall... yes
checking whether C compiler accepts -Wextra... yes
checking whether C compiler accepts -Wdeclaration-after-statement... yes
checking whether C compiler accepts -Wformat... yes
checking whether C compiler accepts -Wformat-security... yes
checking whether C compiler accepts -Wmissing-declarations... yes
checking whether C compiler accepts -Wno-unused-parameter... yes
checking whether C compiler accepts -Wpointer-arith... yes
checking whether C compiler accepts -Wstrict-prototypes... yes
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for mt... no
checking if : is a manifest tool... no
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking for shl_load... no
checking for shl_load in -ldld... no
checking for dlopen... no
checking for dlopen in -ldl... yes
checking whether a program can dlopen itself... yes
checking whether a statically linked program can dlopen itself... yes
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for apxs... /usr/bin/apxs
configure: apxs found at /usr/bin/apxs
checking for apr-1-config... /usr/bin/apr-1-config
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
configure: WARNING: Test coverage requested when not building tests (--with-check=no)
checking for pkg-config... /usr/bin/pkg-config
checking whether compiling and linking against OpenSSL works... yes
checking for CRYPTO_new_ex_data in -lcrypto... yes
checking openssl/crypto.h usability... yes
checking openssl/crypto.h presence... yes
checking for openssl/crypto.h... yes
checking openssl/x509.h usability... yes
checking openssl/x509.h presence... yes
checking for openssl/x509.h... yes
checking openssl/pem.h usability... yes
checking openssl/pem.h presence... yes
checking for openssl/pem.h... yes
checking openssl/ssl.h usability... yes
checking openssl/ssl.h presence... yes
checking for openssl/ssl.h... yes
checking openssl/err.h usability... yes
checking openssl/err.h presence... yes
checking for openssl/err.h... yes
checking for gawk... (cached) gawk
checking for curl-config... /usr/bin/curl-config
checking for the version of libcurl... 7.29.0
checking whether libcurl is usable... yes
checking for curl_free... yes
checking for pcre_compile in -lpcre... yes
checking netdb.h usability... yes
checking netdb.h presence... yes
checking for netdb.h... yes
checking stddef.h usability... yes
checking stddef.h presence... yes
checking for stddef.h... yes
checking sys/socket.h usability... yes
checking sys/socket.h presence... yes
checking for sys/socket.h... yes
checking for unistd.h... (cached) yes
checking for size_t... yes
checking for gethostbyname... yes
checking for memset... yes
checking for socket... yes
checking for strcasecmp... yes
checking for strchr... yes
checking for strncasecmp... yes
checking for strstr... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating src/Makefile
config.status: creating tests/Makefile
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands
config.status: executing libtool commands
make  all-recursive
make[1]: Entering directory `/opt/mod_auth_cas'
Making all in src
make[2]: Entering directory `/opt/mod_auth_cas/src'
make[2]: Nothing to be done for `all'.
make[2]: Leaving directory `/opt/mod_auth_cas/src'
Making all in tests
make[2]: Entering directory `/opt/mod_auth_cas/tests'
make[2]: Nothing to be done for `all'.
make[2]: Leaving directory `/opt/mod_auth_cas/tests'
make[2]: Entering directory `/opt/mod_auth_cas'
make[2]: Leaving directory `/opt/mod_auth_cas'
make[1]: Leaving directory `/opt/mod_auth_cas'
Making install in src
make[1]: Entering directory `/opt/mod_auth_cas/src'
if test ! -d /usr/lib64/httpd/modules; then mkdir -p /usr/lib64/httpd/modules; fi
/usr/bin/apxs -i -S LIBEXECDIR=/usr/lib64/httpd/modules mod_auth_cas.la
/usr/lib64/httpd/build/instdso.sh SH_LIBTOOL='/usr/lib64/apr-1/build/libtool' mod_auth_cas.la /usr/lib64/httpd/modules
/usr/lib64/apr-1/build/libtool --mode=install install mod_auth_cas.la /usr/lib64/httpd/modules/
libtool: install: install .libs/mod_auth_cas.so /usr/lib64/httpd/modules/mod_auth_cas.so
libtool: install: install .libs/mod_auth_cas.lai /usr/lib64/httpd/modules/mod_auth_cas.la
libtool: install: install .libs/mod_auth_cas.a /usr/lib64/httpd/modules/mod_auth_cas.a
libtool: install: chmod 644 /usr/lib64/httpd/modules/mod_auth_cas.a
libtool: install: ranlib /usr/lib64/httpd/modules/mod_auth_cas.a
libtool: finish: PATH="/sbin:/bin:/usr/sbin:/usr/bin:/sbin" ldconfig -n /usr/lib64/httpd/modules
----------------------------------------------------------------------
Libraries have been installed in:
   /usr/lib64/httpd/modules

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the `-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the `LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the `LD_RUN_PATH' environment variable
     during linking
   - use the `-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to `/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
chmod 755 /usr/lib64/httpd/modules/mod_auth_cas.so
make[1]: Leaving directory `/opt/mod_auth_cas/src'
Making install in tests
make[1]: Entering directory `/opt/mod_auth_cas/tests'
make[2]: Entering directory `/opt/mod_auth_cas/tests'
make[2]: Nothing to be done for `install-exec-am'.
make[2]: Nothing to be done for `install-data-am'.
make[2]: Leaving directory `/opt/mod_auth_cas/tests'
make[1]: Leaving directory `/opt/mod_auth_cas/tests'
make[1]: Entering directory `/opt/mod_auth_cas'
make[2]: Entering directory `/opt/mod_auth_cas'
make[2]: Nothing to be done for `install-exec-am'.
make[2]: Nothing to be done for `install-data-am'.
make[2]: Leaving directory `/opt/mod_auth_cas'
make[1]: Leaving directory `/opt/mod_auth_cas'
```

**Comment by dhawes** [comment]
Please "make clean" and then try the make again.

**Comment by agilis1** [comment]
Done.

```
vagrant@localhost /opt/mod_auth_cas (master *=) $ sudo autoreconf -ivf && CPPFLAGS=-I/usr/local/apr/include/apr-1 LDFLAGS=-L/usr/local/apr/lib ./configure && sudo make && sudo make install
autoreconf: Entering directory `.'
autoreconf: configure.ac: not using Gettext
autoreconf: running: aclocal --force -I m4
autoreconf: configure.ac: tracing
autoreconf: running: libtoolize --copy --force
libtoolize: putting auxiliary files in `.'.
libtoolize: copying file `./ltmain.sh'
libtoolize: putting macros in AC_CONFIG_MACRO_DIR, `m4'.
libtoolize: copying file `m4/libtool.m4'
libtoolize: copying file `m4/ltoptions.m4'
libtoolize: copying file `m4/ltsugar.m4'
libtoolize: copying file `m4/ltversion.m4'
libtoolize: copying file `m4/lt~obsolete.m4'
autoreconf: running: /usr/bin/autoconf --force
autoreconf: running: /usr/bin/autoheader --force
autoreconf: running: automake --add-missing --copy --force-missing
autoreconf: Leaving directory `.'
checking whether to enable maintainer-specific portions of Makefiles... no
checking build system type... x86_64-unknown-linux-gnu
checking host system type... x86_64-unknown-linux-gnu
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking how to run the C preprocessor... gcc -E
checking for grep that handles long lines and -e... /usr/bin/grep
checking for egrep... /usr/bin/grep -E
checking for ANSI C header files... yes
checking for sys/types.h... yes
checking for sys/stat.h... yes
checking for stdlib.h... yes
checking for string.h... yes
checking for memory.h... yes
checking for strings.h... yes
checking for inttypes.h... yes
checking for stdint.h... yes
checking for unistd.h... yes
checking minix/config.h usability... no
checking minix/config.h presence... no
checking for minix/config.h... no
checking whether it is safe to define __EXTENSIONS__... yes
checking for library containing strerror... none required
checking for a BSD-compatible install... /usr/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /usr/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking for style of include used by make... GNU
checking whether make supports nested variables... yes
checking dependency style of gcc... gcc3
checking whether make supports nested variables... (cached) yes
checking for gcc... (cached) gcc
checking whether we are using the GNU C compiler... (cached) yes
checking whether gcc accepts -g... (cached) yes
checking for gcc option to accept ISO C89... (cached) none needed
checking whether gcc and cc understand -c and -o together... yes
checking whether C compiler accepts -Wall... yes
checking whether C compiler accepts -Wextra... yes
checking whether C compiler accepts -Wdeclaration-after-statement... yes
checking whether C compiler accepts -Wformat... yes
checking whether C compiler accepts -Wformat-security... yes
checking whether C compiler accepts -Wmissing-declarations... yes
checking whether C compiler accepts -Wno-unused-parameter... yes
checking whether C compiler accepts -Wpointer-arith... yes
checking whether C compiler accepts -Wstrict-prototypes... yes
checking how to print strings... printf
checking for a sed that does not truncate output... /usr/bin/sed
checking for fgrep... /usr/bin/grep -F
checking for ld used by gcc... /usr/bin/ld
checking if the linker (/usr/bin/ld) is GNU ld... yes
checking for BSD- or MS-compatible name lister (nm)... /usr/bin/nm -B
checking the name lister (/usr/bin/nm -B) interface... BSD nm
checking whether ln -s works... yes
checking the maximum length of command line arguments... 1572864
checking whether the shell understands some XSI constructs... yes
checking whether the shell understands "+="... yes
checking how to convert x86_64-unknown-linux-gnu file names to x86_64-unknown-linux-gnu format... func_convert_file_noop
checking how to convert x86_64-unknown-linux-gnu file names to toolchain format... func_convert_file_noop
checking for /usr/bin/ld option to reload object files... -r
checking for objdump... objdump
checking how to recognize dependent libraries... pass_all
checking for dlltool... no
checking how to associate runtime and link libraries... printf %s\n
checking for ar... ar
checking for archiver @FILE support... @
checking for strip... strip
checking for ranlib... ranlib
checking command to parse /usr/bin/nm -B output from gcc object... ok
checking for sysroot... no
checking for mt... no
checking if : is a manifest tool... no
checking for dlfcn.h... yes
checking for objdir... .libs
checking if gcc supports -fno-rtti -fno-exceptions... no
checking for gcc option to produce PIC... -fPIC -DPIC
checking if gcc PIC flag -fPIC -DPIC works... yes
checking if gcc static flag -static works... no
checking if gcc supports -c -o file.o... yes
checking if gcc supports -c -o file.o... (cached) yes
checking whether the gcc linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking whether -lc should be explicitly linked in... no
checking dynamic linker characteristics... GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking for shl_load... no
checking for shl_load in -ldld... no
checking for dlopen... no
checking for dlopen in -ldl... yes
checking whether a program can dlopen itself... yes
checking whether a statically linked program can dlopen itself... yes
checking whether stripping libraries is possible... yes
checking if libtool supports shared libraries... yes
checking whether to build shared libraries... yes
checking whether to build static libraries... yes
checking for apxs... /usr/bin/apxs
configure: apxs found at /usr/bin/apxs
checking for apr-1-config... /usr/bin/apr-1-config
checking for pkg-config... /usr/bin/pkg-config
checking pkg-config is at least version 0.9.0... yes
configure: WARNING: Test coverage requested when not building tests (--with-check=no)
checking for pkg-config... /usr/bin/pkg-config
checking whether compiling and linking against OpenSSL works... yes
checking for CRYPTO_new_ex_data in -lcrypto... yes
checking openssl/crypto.h usability... yes
checking openssl/crypto.h presence... yes
checking for openssl/crypto.h... yes
checking openssl/x509.h usability... yes
checking openssl/x509.h presence... yes
checking for openssl/x509.h... yes
checking openssl/pem.h usability... yes
checking openssl/pem.h presence... yes
checking for openssl/pem.h... yes
checking openssl/ssl.h usability... yes
checking openssl/ssl.h presence... yes
checking for openssl/ssl.h... yes
checking openssl/err.h usability... yes
checking openssl/err.h presence... yes
checking for openssl/err.h... yes
checking for gawk... (cached) gawk
checking for curl-config... /usr/bin/curl-config
checking for the version of libcurl... 7.29.0
checking whether libcurl is usable... yes
checking for curl_free... yes
checking for pcre_compile in -lpcre... yes
checking netdb.h usability... yes
checking netdb.h presence... yes
checking for netdb.h... yes
checking stddef.h usability... yes
checking stddef.h presence... yes
checking for stddef.h... yes
checking sys/socket.h usability... yes
checking sys/socket.h presence... yes
checking for sys/socket.h... yes
checking for unistd.h... (cached) yes
checking for size_t... yes
checking for gethostbyname... yes
checking for memset... yes
checking for socket... yes
checking for strcasecmp... yes
checking for strchr... yes
checking for strncasecmp... yes
checking for strstr... yes
checking that generated files are newer than configure... done
configure: creating ./config.status
config.status: creating Makefile
config.status: creating src/Makefile
config.status: creating tests/Makefile
config.status: creating config.h
config.status: config.h is unchanged
config.status: executing depfiles commands
config.status: executing libtool commands
make  all-recursive
make[1]: Entering directory `/opt/mod_auth_cas'
Making all in src
make[2]: Entering directory `/opt/mod_auth_cas/src'
/usr/bin/apxs -c -lpcre -lcrypto  -Wc,"-I/usr/local/apr/include/apr-1 -D_LARGEFILE64_SOURCE -g -O2 -Wall -Wextra -Wdeclaration-after-statement -Wformat -Wformat-security -Wmissing-declarations -Wno-unused-parameter -Wpointer-arith -Wstrict-prototypes " -Wl," -L/usr/local/apr/lib -lcurl   -lpcre -lcrypto " mod_auth_cas.c cas_saml_attr.c
/usr/lib64/apr-1/build/libtool --silent --mode=compile gcc -std=gnu99 -prefer-pic -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic  -DLINUX -D_REENTRANT -D_GNU_SOURCE -pthread -I/usr/include/httpd  -I/usr/include/apr-1   -I/usr/include/apr-1  -I/usr/local/apr/include/apr-1 -D_LARGEFILE64_SOURCE -g -O2 -Wall -Wextra -Wdeclaration-after-statement -Wformat -Wformat-security -Wmissing-declarations -Wno-unused-parameter -Wpointer-arith -Wstrict-prototypes   -c -o mod_auth_cas.lo mod_auth_cas.c && touch mod_auth_cas.slo
/usr/lib64/apr-1/build/libtool --silent --mode=compile gcc -std=gnu99 -prefer-pic -O2 -g -pipe -Wall -Wp,-D_FORTIFY_SOURCE=2 -fexceptions -fstack-protector-strong --param=ssp-buffer-size=4 -grecord-gcc-switches -m64 -mtune=generic  -DLINUX -D_REENTRANT -D_GNU_SOURCE -pthread -I/usr/include/httpd  -I/usr/include/apr-1   -I/usr/include/apr-1  -I/usr/local/apr/include/apr-1 -D_LARGEFILE64_SOURCE -g -O2 -Wall -Wextra -Wdeclaration-after-statement -Wformat -Wformat-security -Wmissing-declarations -Wno-unused-parameter -Wpointer-arith -Wstrict-prototypes   -c -o cas_saml_attr.lo cas_saml_attr.c && touch cas_saml_attr.slo
/usr/lib64/apr-1/build/libtool --silent --mode=link gcc -std=gnu99 -Wl,-z,relro,-z,now   -o mod_auth_cas.la  -L/usr/local/apr/lib -lcurl   -lpcre -lcrypto   -lpcre -lcrypto -rpath /usr/lib64/httpd/modules -module -avoid-version    cas_saml_attr.lo mod_auth_cas.lo
make[2]: Leaving directory `/opt/mod_auth_cas/src'
Making all in tests
make[2]: Entering directory `/opt/mod_auth_cas/tests'
make[2]: Nothing to be done for `all'.
make[2]: Leaving directory `/opt/mod_auth_cas/tests'
make[2]: Entering directory `/opt/mod_auth_cas'
make[2]: Leaving directory `/opt/mod_auth_cas'
make[1]: Leaving directory `/opt/mod_auth_cas'
Making install in src
make[1]: Entering directory `/opt/mod_auth_cas/src'
if test ! -d /usr/lib64/httpd/modules; then mkdir -p /usr/lib64/httpd/modules; fi
/usr/bin/apxs -i -S LIBEXECDIR=/usr/lib64/httpd/modules mod_auth_cas.la
/usr/lib64/httpd/build/instdso.sh SH_LIBTOOL='/usr/lib64/apr-1/build/libtool' mod_auth_cas.la /usr/lib64/httpd/modules
/usr/lib64/apr-1/build/libtool --mode=install install mod_auth_cas.la /usr/lib64/httpd/modules/
libtool: install: install .libs/mod_auth_cas.so /usr/lib64/httpd/modules/mod_auth_cas.so
libtool: install: install .libs/mod_auth_cas.lai /usr/lib64/httpd/modules/mod_auth_cas.la
libtool: install: install .libs/mod_auth_cas.a /usr/lib64/httpd/modules/mod_auth_cas.a
libtool: install: chmod 644 /usr/lib64/httpd/modules/mod_auth_cas.a
libtool: install: ranlib /usr/lib64/httpd/modules/mod_auth_cas.a
libtool: finish: PATH="/sbin:/bin:/usr/sbin:/usr/bin:/sbin" ldconfig -n /usr/lib64/httpd/modules
----------------------------------------------------------------------
Libraries have been installed in:
   /usr/lib64/httpd/modules

If you ever happen to want to link against installed libraries
in a given directory, LIBDIR, you must either use libtool, and
specify the full pathname of the library, or use the `-LLIBDIR'
flag during linking and do at least one of the following:
   - add LIBDIR to the `LD_LIBRARY_PATH' environment variable
     during execution
   - add LIBDIR to the `LD_RUN_PATH' environment variable
     during linking
   - use the `-Wl,-rpath -Wl,LIBDIR' linker flag
   - have your system administrator add LIBDIR to `/etc/ld.so.conf'

See any operating system documentation about shared libraries for
more information, such as the ld(1) and ld.so(8) manual pages.
----------------------------------------------------------------------
chmod 755 /usr/lib64/httpd/modules/mod_auth_cas.so
make[1]: Leaving directory `/opt/mod_auth_cas/src'
Making install in tests
make[1]: Entering directory `/opt/mod_auth_cas/tests'
make[2]: Entering directory `/opt/mod_auth_cas/tests'
make[2]: Nothing to be done for `install-exec-am'.
make[2]: Nothing to be done for `install-data-am'.
make[2]: Leaving directory `/opt/mod_auth_cas/tests'
make[1]: Leaving directory `/opt/mod_auth_cas/tests'
make[1]: Entering directory `/opt/mod_auth_cas'
make[2]: Entering directory `/opt/mod_auth_cas'
make[2]: Nothing to be done for `install-exec-am'.
make[2]: Nothing to be done for `install-data-am'.
make[2]: Leaving directory `/opt/mod_auth_cas'
make[1]: Leaving directory `/opt/mod_auth_cas'
```

**Comment by agilis1** [comment]
When I perform the following command I now get this:

```
nm /etc/httpd/modules/mod_auth_cas.so | grep cas_attr_builder_new
0000000000004590 T cas_attr_builder_new
```

That seems like progress!

**Comment by agilis1** [comment]
I think that did it! I just needed to do the following:

`sudo mkdir /var/cache/httpd/mod_auth_cas/`
`sudo chmod -R 700 /var/cache/httpd/mod_auth_cas`

Apache now starts and it shows the auto_cas_module is now loaded properly:

```
[root@localhost httpd]# sudo httpd -M | grep auth_cas_module
AH00558: httpd: Could not reliably determine the server's fully qualified domain name, using localhost.localdomain. Set the 'ServerName' directive globally to suppress this message
 auth_cas_module (shared)

```

I admit I don't have a tremendous amount of experience with compiling packages from source. 

Thank you very much for your time and help. I greatly appreciate it!! Take care.

**Comment by dhawes** [comment]
I'm glad recompiling worked! I admit I wasn't looking forward to figuring out how the module was compiled without those symbols. ;)

**Comment by agilis1** [comment]
Haha I have no idea either. That should never have happened. My guess is during my trial and error. At one point I tried with v1.5.2 of Apache Portable Runtime and ran into the issue where I could not get httpd to load with APR v1.5.2 so trying to research how to address that issue I found your instructions in a previous issue using v1.7. Unfortunately I had the same issue.

So my guess during that whole phase something got really messed up.
